File: NeteaseM2DClient/app/src/main/java/com/aoihosizora/neteasem2dclient/MainService.java
package com.aoihosizora.neteasem2dclient;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.media.MediaMetadata;
import android.media.session.MediaController;
import android.media.session.MediaSessionManager;
import android.media.session.PlaybackState;
import android.os.IBinder;
import android.service.notification.NotificationListenerService;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.widget.Toast;

import com.aoihosizora.neteasem2dclient.model.DestroyDto;
import com.aoihosizora.neteasem2dclient.model.MetadataDto;
import com.aoihosizora.neteasem2dclient.model.PlaybackStateDto;

import org.json.JSONObject;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;

public class MainService extends NotificationListenerService {

    /**
     * 外部事件，由 MainActivity 初始化
     */
    public static MainEvent mainEvent = null;

    /**
     * 媒体控制器
     */
    public static MediaController mediaController = null;

    /**
     * 媒体回调
     */
    public static MediaCallBack mediaCallBack = null;

    @Override
    public void onCreate() {
        super.onCreate();
        if (mediaController != null) {
            return;
        }
        if (mediaCallBack == null) {
            mediaCallBack = new MediaCallBack();
        }

        try {
            // 获得 MediaSessionManager
            MediaSessionManager manager = (MediaSessionManager) getSystemService(Context.MEDIA_SESSION_SERVICE);
            if (manager == null) {
                if (mainEvent != null) {
                    mainEvent.onNoSession();
                }
                return;
            }

            // 检查当前服务是否是网易云
            List<MediaController> controllers = manager.getActiveSessions(new ComponentName(this, MainService.class));
            for (MediaController controller : controllers) {
                String packageName = controller.getPackageName();
                if (packageName.contains("netease") && packageName.contains("music")) {
                    mediaController = controller;
                    break;
                }
            }
            if (mediaController == null) {
                if (mainEvent != null) {
                    mainEvent.onNoSession();
                }
                return;
            }

            // 注册回调
            mediaController.registerCallback(mediaCallBack);

            // 返回状态
            PlaybackState state = mediaController.getPlaybackState();
            MediaMetadata metadata = mediaController.getMetadata();
            if (state != null && metadata != null) {
                if (state.getState() == PlaybackState.STATE_PLAYING) {
                    mediaCallBack.onMetadataChanged(mediaController.getMetadata());
                    mediaCallBack.onPlaybackStateChanged(mediaController.getPlaybackState());
                }
            }
        } catch (Exception ex) {
            Toast.makeText(this, getExceptionString(ex), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return super.onBind(intent);
    }

    /**
     * 重写媒体回调
     */
    class MediaCallBack extends MediaController.Callback {

        /**
         * 播放状态
         */
        @Override
        public void onPlaybackStateChanged(@Nullable PlaybackState state) {
            super.onPlaybackStateChanged(state);
            // Toast.makeText(MainService.this, "onPlaybackStateChanged", Toast.LENGTH_SHORT).show();
            if (state == null) {
                return;
            }

            // 信息
            boolean isPlaying = state.getState() == PlaybackState.STATE_PLAYING;
            double musicPosition = (double) state.getPosition() / 1000.0;

            // 序列化
            PlaybackStateDto dto = new PlaybackStateDto(isPlaying, musicPosition);
            JSONObject json = dto.toJson();
            if (json == null) {
                return;
            }
            Toast.makeText(MainService.this, json.toString(), Toast.LENGTH_SHORT).show();

            // 发送
            new Thread(() -> {
                boolean ok = SendUtils.send(json.toString());
                if (!ok && mainEvent != null) {
                    mainEvent.onDisconnect();
                }
            }).start();
        }

        /**
         * 媒体信息
         */
        @Override
        public void onMetadataChanged(@Nullable MediaMetadata metadata) {
            super.onMetadataChanged(metadata);
            // Toast.makeText(MainService.this, "onMetadataChanged", Toast.LENGTH_SHORT).show();
            if (metadata == null) {
                return;
            }

            // 信息
            String title = metadata.getString(MediaMetadata.METADATA_KEY_TITLE);
            String artist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST);
            String album = metadata.getString(MediaMetadata.METADATA_KEY_ALBUM);
            double duration = (double) metadata.getLong(MediaMetadata.METADATA_KEY_DURATION) / 1000.0;

            // 序列化
            MetadataDto dto = new MetadataDto(title, artist, album, duration);
            JSONObject json = dto.toJson();
            if (json == null) {
                return;
            }
            Toast.makeText(MainService.this, json.toString(), Toast.LENGTH_SHORT).show();

            // 发送
            new Thread(() -> {
                boolean ok = SendUtils.send(json.toString());
                if (!ok && mainEvent != null) {
                    mainEvent.onDisconnect();
                }
            }).start();
        }

        /**
         * 退出
         */
        @Override
        public void onSessionDestroyed() {
            super.onSessionDestroyed();
            Toast.makeText(MainService.this, "onSessionDestroyed", Toast.LENGTH_SHORT).show();
            if (mainEvent != null) {
                mainEvent.onSessionDestroy();
            }

            // 序列化
            DestroyDto dto = new DestroyDto(true);
            JSONObject json = dto.toJson();
            if (json == null) {
                return;
            }
            Toast.makeText(MainService.this, json.toString(), Toast.LENGTH_SHORT).show();

            // 发送
            new Thread(() -> {
                boolean ok = SendUtils.send(json.toString());
                if (!ok && mainEvent != null) {
                    mainEvent.onDisconnect();
                }
            }).start();
        }
    }

    /**
     * Debug 用
     */
    @UiThread
    private String getExceptionString(Exception ex) {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        ex.printStackTrace(pw);
        pw.flush();
        return sw.toString();
    }
}

/**
 * 外部事件，没有使用广播
 */
interface MainEvent {

    /**
     * Socket 断开连接
     */
    @UiThread
    void onDisconnect();

    /**
     * 没有网易云会话
     */
    @UiThread
    void onNoSession();

    /**
     * 网易云会话关闭
     */
    @UiThread
    void onSessionDestroy();
}

File: android-client/app/src/main/java/com/aoihosizora/ncmlatwclient/MainService.kt
package com.aoihosizora.ncmlatwclient

import android.app.Activity
import android.app.ActivityManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.media.MediaMetadata
import android.media.session.MediaController
import android.media.session.MediaSessionManager
import android.media.session.PlaybackState
import android.service.notification.NotificationListenerService
import androidx.annotation.UiThread

class MainService : NotificationListenerService() {

    companion object {
        var mediaController: MediaController? = null
        var mediaCallback: MediaCallBack? = null
        var eventListener: EventListener? = null

        fun isRunning(a: Activity): Boolean {
            val manager = a.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            for (service in manager.getRunningServices(Int.MAX_VALUE)) {
                if (service.service.className == MainService::class.java.name) {
                    return true
                }
            }
            return false
        }
    }

    // ========================
    // service lifetime related
    // ========================

    override fun onCreate() {
        super.onCreate()
        eventListener?.onServiceLifetime("onCreate")

        if (mediaController != null) {
            stopSelf()
            return
        }
        if (mediaCallback == null) {
            mediaCallback = MediaCallBack()
        }

        try {
            // get MediaSessionManager
            val manager = getSystemService(Context.MEDIA_SESSION_SERVICE) as? MediaSessionManager
            if (manager == null) {
                eventListener?.onNoSession()
                stopSelf()
                return
            }
            // find ncm MediaController
            for (controller in manager.getActiveSessions(ComponentName(this, MainService::class.java))) {
                val pkgName = controller.packageName
                if (pkgName.contains("netease") && pkgName.contains("music")) {
                    mediaController = controller
                    break
                }
            }
            if (mediaController == null) {
                eventListener?.onNoSession()
                stopSelf()
                return
            }

            // register callback
            mediaController!!.registerCallback(mediaCallback!!)
        } catch (ex: Exception) {
            // Toast.makeText(this, getExceptionString(ex), Toast.LENGTH_SHORT).show()
            stopSelf()
        }

        // get current states
        val state = mediaController!!.playbackState
        val metadata = mediaController!!.metadata
        if (state != null && metadata != null) {
            mediaCallback!!.onMetadataChanged(metadata)
            mediaCallback!!.onPlaybackStateChanged(state)
        }
    }

    override fun onDestroy() {
        eventListener?.onServiceLifetime("onDestroy")
        super.onDestroy()
    }

    // override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    //     intent?.let { it ->
    //         if (it.action == "UPDATE") {
    //             mediaController?.let { it1 ->
    //                 val state = it1.playbackState
    //                 val metadata = it1.metadata
    //                 if (state != null && metadata != null) {
    //                     val isPlaying = state.state == PlaybackState.STATE_PLAYING
    //                     val position = state.position.toDouble() / 1000.0
    //                     val title = metadata.getString(MediaMetadata.METADATA_KEY_TITLE)
    //                     val artist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST)
    //                     val album = metadata.getString(MediaMetadata.METADATA_KEY_ALBUM)
    //                     val duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION).toDouble() / 1000.0
    //                     Thread {
    //                         PlaybackStateDto(isPlaying, position).toJSON()?.let { SocketUtils.send(it.toString()) }
    //                         MetadataDto(title, artist, album, duration).toJSON()?.let { SocketUtils.send(it.toString()) }
    //                     }.start()
    //                 }
    //             }
    //         }
    //     }
    //     return START_STICKY
    // }

    // ======================
    // media callback related
    // ======================

    inner class MediaCallBack : MediaController.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackState?) {
            super.onPlaybackStateChanged(state)
            if (state == null) {
                return
            }

            // information
            val isPlaying = state.state == PlaybackState.STATE_PLAYING
            val position = state.position.toDouble() / 1000.0
            eventListener?.onPlaybackStateChanged(isPlaying, position)
        }

        override fun onMetadataChanged(metadata: MediaMetadata?) {
            super.onMetadataChanged(metadata)
            if (metadata == null) {
                return
            }

            // information
            val title = metadata.getString(MediaMetadata.METADATA_KEY_TITLE)
            val artist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST)
            val album = metadata.getString(MediaMetadata.METADATA_KEY_ALBUM)
            val duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION).toDouble() / 1000.0
            eventListener?.onMetadataChanged(title, artist, album, duration)
        }

        override fun onSessionDestroyed() {
            super.onSessionDestroyed()
            eventListener?.onSessionDestroyed()
            stopSelf()
        }
    }

    interface EventListener {
        /**
         * No ncm session callback.
         */
        @UiThread
        fun onNoSession()

        /**
         * Session destroyed callback.
         */
        @UiThread
        fun onSessionDestroyed()

        /**
         * Media playback state changed callback.
         */
        @UiThread
        fun onPlaybackStateChanged(isPlaying: Boolean, position: Double)

        /**
         * Media metadata changed callback.
         */
        @UiThread
        fun onMetadataChanged(title: String, artist: String, album: String, duration: Double)

        /**
         * Service created or destroyed callback.
         */
        @UiThread
        fun onServiceLifetime(text: String)
    }

    // private fun getExceptionString(ex: Exception): String? {
    //     val sw = StringWriter()
    //     val pw = PrintWriter(sw)
    //     ex.printStackTrace(pw)
    //     pw.flush()
    //     return sw.toString()
    // }
}
